File: project files/server/src/services/adminImportService.js
const catalogAdminService = require('./catalogAdminService');
const importJobRepository = require('../repositories/importJobRepository');
const userRepository = require('../repositories/userRepository');
const { sendEmail } = require('./emailService');

const PRODUCT_CSV_TEMPLATE_HEADERS = [
  'name',
  'slug',
  'SKU',
  'brand',
  'categoryId',
  'subCategoryId',
  'description',
  'weight',
  'price',
  'MRP',
  'stock',
  'isActive',
  'isFeatured',
  'sellerId'
];

const STOCK_CSV_TEMPLATE_HEADERS = ['productId', 'variantId', 'stock'];

const parseCsv = (csvContent) => {
  const lines = String(csvContent || '')
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean);
  if (!lines.length) return { headers: [], rows: [] };
  const headers = lines[0].split(',').map((entry) => entry.trim());
  const rows = lines.slice(1).map((line) => line.split(',').map((entry) => entry.trim()));
  return { headers, rows };
};

const parseBoolean = (value, defaultValue = false) => {
  if (value === undefined || value === null || value === '') return defaultValue;
  const normalized = String(value).toLowerCase();
  return ['1', 'true', 'yes', 'y'].includes(normalized);
};

const processProductCsvImport = async ({ jobId, csvContent, actor }) => {
  const job = await importJobRepository.findById(jobId);
  if (!job) return;

  job.status = 'RUNNING';
  job.startedAt = new Date();
  await importJobRepository.save(job);

  try {
    const parsed = parseCsv(csvContent);
    const { headers, rows } = parsed;
    const normalizedHeaders = headers.map((header) => header.trim());
    job.totalRows = rows.length;
    await importJobRepository.save(job);

    for (let index = 0; index < rows.length; index += 1) {
      const row = rows[index];
      const rowNumber = index + 2;
      try {
        const values = {};
        normalizedHeaders.forEach((header, headerIndex) => {
          values[header] = row[headerIndex];
        });

        if (
          !values.name ||
          !values.slug ||
          !values.SKU ||
          !values.brand ||
          !values.categoryId ||
          !values.subCategoryId ||
          !values.weight
        ) {
          throw new Error('Missing required columns');
        }

        const price = Number(values.price);
        const mrp = Number(values.MRP);
        const stock = Number(values.stock);
        if (!Number.isFinite(price) || !Number.isFinite(mrp) || !Number.isFinite(stock)) {
          throw new Error('Invalid numeric fields');
        }

        const payload = {
          name: values.name,
          slug: values.slug,
          SKU: values.SKU,
          brand: values.brand,
          categoryId: values.categoryId,
          subCategoryId: values.subCategoryId,
          sellerId: values.sellerId || actor.userId,
          description: values.description || '',
          isActive: parseBoolean(values.isActive, true),
          isFeatured: parseBoolean(values.isFeatured, false),
          images: ['https://placehold.co/800x800?text=Product'],
          tags: [values.brand],
          variants: [
            {
              weight: values.weight,
              price,
              MRP: mrp,
              stock,
              skuSuffix: `${values.SKU}-1`
            }
          ]
        };
        await catalogAdminService.upsertProduct(payload, actor);
        job.successRows += 1;
      } catch (error) {
        job.failedRows += 1;
        job.failureReport.push({ rowNumber, reason: error.message || 'Import failed' });
      } finally {
        job.processedRows += 1;
        if (job.processedRows % 10 === 0 || job.processedRows === job.totalRows) {
          await importJobRepository.save(job);
        }
      }
    }

    job.status = 'COMPLETED';
    job.completedAt = new Date();
    await importJobRepository.save(job);
  } catch (error) {
    job.status = 'FAILED';
    job.errorMessage = error.message || 'Import failed';
    job.completedAt = new Date();
    await importJobRepository.save(job);
  }

  const user = await userRepository.findById(actor.userId);
  if (user?.email) {
    await sendEmail({
      to: user.email,
      subject: `Product CSV import ${job.status.toLowerCase()}`,
      text: `Import job ${job._id} finished with status ${job.status}. Success: ${job.successRows}, Failed: ${job.failedRows}.`
    });
  }
  const finalDoc = await importJobRepository.findById(job._id);
  if (finalDoc) {
    finalDoc.notificationSent = true;
    await importJobRepository.save(finalDoc);
  }
};

const createProductCsvImportJob = async ({ csvContent, actor }) => {
  const job = await importJobRepository.createJob({
    jobType: 'PRODUCT_CSV',
    status: 'PENDING',
    createdBy: actor.userId
  });
  setImmediate(() => {
    processProductCsvImport({ jobId: job._id, csvContent, actor }).catch(() => {});
  });
  return job;
};

const processStockCsvImport = async ({ jobId, csvContent, actor }) => {
  const job = await importJobRepository.findById(jobId);
  if (!job) return;

  job.status = 'RUNNING';
  job.startedAt = new Date();
  await importJobRepository.save(job);

  try {
    const parsed = parseCsv(csvContent);
    const { headers, rows } = parsed;
    const normalizedHeaders = headers.map((header) => header.trim());
    job.totalRows = rows.length;
    await importJobRepository.save(job);

    for (let index = 0; index < rows.length; index += 1) {
      const row = rows[index];
      const rowNumber = index + 2;
      try {
        const values = {};
        normalizedHeaders.forEach((header, headerIndex) => {
          values[header] = row[headerIndex];
        });
        if (!values.productId || !values.variantId) {
          throw new Error('Missing productId/variantId');
        }
        const stock = Number(values.stock);
        if (!Number.isFinite(stock) || stock < 0) throw new Error('Invalid stock');
        await catalogAdminService.updateVariantStock({
          productId: values.productId,
          variantId: values.variantId,
          stock,
          performedBy: actor.userId,
          reason: 'ADMIN_BULK_STOCK_UPLOAD'
        });
        job.successRows += 1;
      } catch (error) {
        job.failedRows += 1;
        job.failureReport.push({ rowNumber, reason: error.message || 'Import failed' });
      } finally {
        job.processedRows += 1;
        if (job.processedRows % 10 === 0 || job.processedRows === job.totalRows) {
          await importJobRepository.save(job);
        }
      }
    }
    job.status = 'COMPLETED';
    job.completedAt = new Date();
    await importJobRepository.save(job);
  } catch (error) {
    job.status = 'FAILED';
    job.errorMessage = error.message || 'Import failed';
    job.completedAt = new Date();
    await importJobRepository.save(job);
  }

  const user = await userRepository.findById(actor.userId);
  if (user?.email) {
    await sendEmail({
      to: user.email,
      subject: `Stock CSV import ${job.status.toLowerCase()}`,
      text: `Import job ${job._id} finished with status ${job.status}. Success: ${job.successRows}, Failed: ${job.failedRows}.`
    });
  }
  const finalDoc = await importJobRepository.findById(job._id);
  if (finalDoc) {
    finalDoc.notificationSent = true;
    await importJobRepository.save(finalDoc);
  }
};

const createStockCsvImportJob = async ({ csvContent, actor }) => {
  const job = await importJobRepository.createJob({
    jobType: 'STOCK_CSV',
    status: 'PENDING',
    createdBy: actor.userId
  });
  setImmediate(() => {
    processStockCsvImport({ jobId: job._id, csvContent, actor }).catch(() => {});
  });
  return job;
};

const getImportJobStatus = (jobId) => importJobRepository.findByIdLean(jobId);

const getProductCsvTemplate = () => `${PRODUCT_CSV_TEMPLATE_HEADERS.join(',')}\n`;
const getStockCsvTemplate = () => `${STOCK_CSV_TEMPLATE_HEADERS.join(',')}\n`;

module.exports = {
  createProductCsvImportJob,
  createStockCsvImportJob,
  getImportJobStatus,
  getProductCsvTemplate,
  getStockCsvTemplate
};

File: spring_back/src/main/java/com/groceryapp/service/AdminImportService.java
package com.groceryapp.service;

import com.groceryapp.model.ImportJob;
import com.groceryapp.model.User;
import com.groceryapp.repository.ImportJobRepository;
import com.groceryapp.repository.UserRepository;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.CompletableFuture;

@Service
public class AdminImportService {
    private static final List<String> PRODUCT_CSV_TEMPLATE_HEADERS = List.of(
            "name", "slug", "SKU", "brand", "categoryId", "subCategoryId", "description",
            "weight", "price", "MRP", "stock", "isActive", "isFeatured", "sellerId"
    );
    private static final List<String> STOCK_CSV_TEMPLATE_HEADERS = List.of("productId", "variantId", "stock");

    private final ImportJobRepository importJobRepository;
    private final UserRepository userRepository;
    private final CatalogAdminService catalogAdminService;
    private final EmailService emailService;

    public AdminImportService(
            ImportJobRepository importJobRepository,
            UserRepository userRepository,
            CatalogAdminService catalogAdminService,
            EmailService emailService
    ) {
        this.importJobRepository = importJobRepository;
        this.userRepository = userRepository;
        this.catalogAdminService = catalogAdminService;
        this.emailService = emailService;
    }

    public ImportJob createProductCsvImportJob(String csvContent, String actorUserId) {
        Instant now = Instant.now();
        ImportJob job = ImportJob.builder()
                .jobType("PRODUCT_CSV")
                .status("PENDING")
                .createdBy(actorUserId)
                .totalRows(0)
                .processedRows(0)
                .successRows(0)
                .failedRows(0)
                .failureReport(new ArrayList<>())
                .notificationSent(false)
                .errorMessage("")
                .createdAt(now)
                .updatedAt(now)
                .build();
        job = importJobRepository.save(job);
        final String jobId = job.getId();
        CompletableFuture.runAsync(() -> processProductCsvImport(jobId, csvContent, actorUserId));
        return job;
    }

    public ImportJob createStockCsvImportJob(String csvContent, String actorUserId) {
        Instant now = Instant.now();
        ImportJob job = ImportJob.builder()
                .jobType("STOCK_CSV")
                .status("PENDING")
                .createdBy(actorUserId)
                .totalRows(0)
                .processedRows(0)
                .successRows(0)
                .failedRows(0)
                .failureReport(new ArrayList<>())
                .notificationSent(false)
                .errorMessage("")
                .createdAt(now)
                .updatedAt(now)
                .build();
        job = importJobRepository.save(job);
        final String jobId = job.getId();
        CompletableFuture.runAsync(() -> processStockCsvImport(jobId, csvContent, actorUserId));
        return job;
    }

    public ImportJob getImportJobStatus(String jobId) {
        return importJobRepository.findById(jobId).orElse(null);
    }

    public String getProductCsvTemplate() {
        return String.join(",", PRODUCT_CSV_TEMPLATE_HEADERS) + "\n";
    }

    public String getStockCsvTemplate() {
        return String.join(",", STOCK_CSV_TEMPLATE_HEADERS) + "\n";
    }

    private void processProductCsvImport(String jobId, String csvContent, String actorUserId) {
        ImportJob job = importJobRepository.findById(jobId).orElse(null);
        if (job == null) return;

        job.setStatus("RUNNING");
        job.setStartedAt(Instant.now());
        job.setUpdatedAt(Instant.now());
        importJobRepository.save(job);

        try {
            CsvData parsed = parseCsv(csvContent);
            List<String> headers = parsed.headers();
            List<List<String>> rows = parsed.rows();
            job.setTotalRows(rows.size());
            importJobRepository.save(job);

            for (int i = 0; i < rows.size(); i++) {
                int rowNumber = i + 2;
                try {
                    Map<String, String> values = rowToMap(headers, rows.get(i));
                    if (blank(values.get("name")) || blank(values.get("slug")) || blank(values.get("SKU"))
                            || blank(values.get("brand")) || blank(values.get("categoryId"))
                            || blank(values.get("subCategoryId")) || blank(values.get("weight"))) {
                        throw new IllegalArgumentException("Missing required columns");
                    }
                    double price = parseDouble(values.get("price"));
                    double mrp = parseDouble(values.get("MRP"));
                    int stock = (int) parseDouble(values.get("stock"));

                    Map<String, Object> payload = new HashMap<>();
                    payload.put("name", values.get("name"));
                    payload.put("slug", values.get("slug"));
                    payload.put("SKU", values.get("SKU"));
                    payload.put("brand", values.get("brand"));
                    payload.put("categoryId", values.get("categoryId"));
                    payload.put("subCategoryId", values.get("subCategoryId"));
                    payload.put("sellerId", blank(values.get("sellerId")) ? actorUserId : values.get("sellerId"));
                    payload.put("description", values.getOrDefault("description", ""));
                    payload.put("isActive", parseBoolean(values.get("isActive"), true));
                    payload.put("isFeatured", parseBoolean(values.get("isFeatured"), false));
                    payload.put("images", List.of("https://placehold.co/800x800?text=Product"));
                    payload.put("tags", List.of(values.get("brand")));
                    payload.put("variants", List.of(Map.of(
                            "weight", values.get("weight"),
                            "price", price,
                            "MRP", mrp,
                            "stock", stock,
                            "skuSuffix", values.get("SKU") + "-1"
                    )));
                    catalogAdminService.upsertProduct(payload, actorUserId);
                    job.setSuccessRows(job.getSuccessRows() + 1);
                } catch (Exception ex) {
                    job.setFailedRows(job.getFailedRows() + 1);
                    List<ImportJob.Failure> failures = job.getFailureReport() == null ? new ArrayList<>() : job.getFailureReport();
                    failures.add(ImportJob.Failure.builder()
                            .rowNumber(rowNumber)
                            .reason(msg(ex))
                            .build());
                    job.setFailureReport(failures);
                } finally {
                    job.setProcessedRows(job.getProcessedRows() + 1);
                    if (job.getProcessedRows() % 10 == 0 || Objects.equals(job.getProcessedRows(), job.getTotalRows())) {
                        job.setUpdatedAt(Instant.now());
                        importJobRepository.save(job);
                    }
                }
            }

            job.setStatus("COMPLETED");
            job.setCompletedAt(Instant.now());
            job.setUpdatedAt(Instant.now());
            importJobRepository.save(job);
        } catch (Exception ex) {
            job.setStatus("FAILED");
            job.setErrorMessage(msg(ex));
            job.setCompletedAt(Instant.now());
            job.setUpdatedAt(Instant.now());
            importJobRepository.save(job);
        }

        notifyJobStatus(job, actorUserId, "Product CSV import");
    }

    private void processStockCsvImport(String jobId, String csvContent, String actorUserId) {
        ImportJob job = importJobRepository.findById(jobId).orElse(null);
        if (job == null) return;

        job.setStatus("RUNNING");
        job.setStartedAt(Instant.now());
        job.setUpdatedAt(Instant.now());
        importJobRepository.save(job);

        try {
            CsvData parsed = parseCsv(csvContent);
            List<String> headers = parsed.headers();
            List<List<String>> rows = parsed.rows();
            job.setTotalRows(rows.size());
            importJobRepository.save(job);

            for (int i = 0; i < rows.size(); i++) {
                int rowNumber = i + 2;
                try {
                    Map<String, String> values = rowToMap(headers, rows.get(i));
                    if (blank(values.get("productId")) || blank(values.get("variantId"))) {
                        throw new IllegalArgumentException("Missing productId/variantId");
                    }
                    int stock = (int) parseDouble(values.get("stock"));
                    if (stock < 0) {
                        throw new IllegalArgumentException("Invalid stock");
                    }
                    catalogAdminService.updateVariantStock(
                            values.get("productId"),
                            values.get("variantId"),
                            stock,
                            actorUserId,
                            "ADMIN_BULK_STOCK_UPLOAD"
                    );
                    job.setSuccessRows(job.getSuccessRows() + 1);
                } catch (Exception ex) {
                    job.setFailedRows(job.getFailedRows() + 1);
                    List<ImportJob.Failure> failures = job.getFailureReport() == null ? new ArrayList<>() : job.getFailureReport();
                    failures.add(ImportJob.Failure.builder()
                            .rowNumber(rowNumber)
                            .reason(msg(ex))
                            .build());
                    job.setFailureReport(failures);
                } finally {
                    job.setProcessedRows(job.getProcessedRows() + 1);
                    if (job.getProcessedRows() % 10 == 0 || Objects.equals(job.getProcessedRows(), job.getTotalRows())) {
                        job.setUpdatedAt(Instant.now());
                        importJobRepository.save(job);
                    }
                }
            }
            job.setStatus("COMPLETED");
            job.setCompletedAt(Instant.now());
            job.setUpdatedAt(Instant.now());
            importJobRepository.save(job);
        } catch (Exception ex) {
            job.setStatus("FAILED");
            job.setErrorMessage(msg(ex));
            job.setCompletedAt(Instant.now());
            job.setUpdatedAt(Instant.now());
            importJobRepository.save(job);
        }

        notifyJobStatus(job, actorUserId, "Stock CSV import");
    }

    private void notifyJobStatus(ImportJob job, String actorUserId, String prefix) {
        User user = userRepository.findById(actorUserId).orElse(null);
        if (user != null && !blank(user.getEmail())) {
            emailService.sendEmail(
                    user.getEmail(),
                    prefix + " " + str(job.getStatus()).toLowerCase(Locale.ROOT),
                    "Import job " + job.getId() + " finished with status " + job.getStatus()
                            + ". Success: " + job.getSuccessRows() + ", Failed: " + job.getFailedRows() + ".",
                    null
            );
        }
        ImportJob saved = importJobRepository.findById(job.getId()).orElse(null);
        if (saved != null) {
            saved.setNotificationSent(true);
            saved.setUpdatedAt(Instant.now());
            importJobRepository.save(saved);
        }
    }

    private CsvData parseCsv(String csvContent) {
        List<String> lines = Arrays.stream(str(csvContent).split("\\r?\\n"))
                .map(String::trim)
                .filter(line -> !line.isBlank())
                .toList();
        if (lines.isEmpty()) {
            return new CsvData(List.of(), List.of());
        }
        List<String> headers = Arrays.stream(lines.get(0).split(","))
                .map(String::trim)
                .toList();
        List<List<String>> rows = new ArrayList<>();
        for (int i = 1; i < lines.size(); i++) {
            rows.add(Arrays.stream(lines.get(i).split(","))
                    .map(String::trim)
                    .toList());
        }
        return new CsvData(headers, rows);
    }

    private Map<String, String> rowToMap(List<String> headers, List<String> row) {
        Map<String, String> values = new HashMap<>();
        for (int i = 0; i < headers.size(); i++) {
            values.put(headers.get(i), i < row.size() ? row.get(i) : "");
        }
        return values;
    }

    private boolean parseBoolean(String value, boolean defaultValue) {
        if (value == null || value.isBlank()) return defaultValue;
        String normalized = value.trim().toLowerCase(Locale.ROOT);
        return List.of("1", "true", "yes", "y").contains(normalized);
    }

    private double parseDouble(String raw) {
        try {
            return Double.parseDouble(str(raw));
        } catch (Exception ex) {
            throw new IllegalArgumentException("Invalid numeric fields");
        }
    }

    private boolean blank(String value) {
        return value == null || value.trim().isBlank();
    }

    private String str(Object value) {
        return value == null ? "" : String.valueOf(value).trim();
    }

    private String msg(Exception ex) {
        return ex == null || ex.getMessage() == null || ex.getMessage().isBlank() ? "Import failed" : ex.getMessage();
    }

    private record CsvData(List<String> headers, List<List<String>> rows) {
    }
}
